### PR TITLE
Replace cnf structhash with custom one

### PR DESCRIPTION
**Pull request by krhubert**
This version has few improvements:

- the most important thing - it only dumps struct fields that **have** hash:name tag (it allows us to skip all XXX_ generated fields by protobuf)
- cover edge cases for dump maps/interface
- 100% test coverage 
- better performance

```
➜  structhash git:(feature/new-structhash) ✗ benchcmp old.out new.out
benchmark                 old ns/op     new ns/op     delta
BenchmarkSimpleJSON-6     852           894           +4.93%
BenchmarkSimpleDump-6     3806          1145          -69.92%
BenchmarkFullJSON-6       5888          6110          +3.77%
BenchmarkFullDump-6       20447         1168          -94.29%
BenchmarkTagsJSON-6       572           594           +3.85%
BenchmarkTagsDump-6       3433          4489          +30.76%

benchmark                 old allocs     new allocs     delta
BenchmarkSimpleJSON-6     1              1              +0.00%
BenchmarkSimpleDump-6     22             12             -45.45%
BenchmarkFullJSON-6       9              9              +0.00%
BenchmarkFullDump-6       130            12             -90.77%
BenchmarkTagsJSON-6       1              1              +0.00%
BenchmarkTagsDump-6       21             34             +61.90%

benchmark                 old bytes     new bytes     delta
BenchmarkSimpleJSON-6     96            96            +0.00%
BenchmarkSimpleDump-6     1056          216           -79.55%
BenchmarkFullJSON-6       960           960           +0.00%
BenchmarkFullDump-6       6336          216           -96.59%
BenchmarkTagsJSON-6       48            48            +0.00%
BenchmarkTagsDump-6       680           960           +41.18%
```

### CODE REVIEW

File: hash/structhash/structhash_test.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+			}{0, 1},
+			"{a:1,b:0}",
+		},
+		{
+			struct {
+				a int `hash:"name:b,omitempty"`
+				b int `hash:"name:a,omitempty"`
+			}{0, 1},
+			"{a:1}",
+		},
```
Comment (NicolasMahe) [inline hash/structhash/structhash_test.go]: there is no test with with empty tag and `hash:"-"`.
Can you add:
```suggestion
		},
		{
			struct {
				a int
				b int `hash:"-"`
				c int `hash:"name:c"`
			}{1, 2, 3},
			"{c:3}",
		},
```

Comment (krhubert) [reply hash/structhash/structhash_test.go]: There is on at line [62](https://github.com/mesg-foundation/engine/pull/1298/files#diff-08fe067a6084b7b01dc3bc035976b2abR62) for no name and line [127](https://github.com/mesg-foundation/engine/pull/1298/files#diff-08fe067a6084b7b01dc3bc035976b2abR127) for `hash:"-"`

Comment (NicolasMahe) [reply hash/structhash/structhash_test.go]: and empty/no tag? `
```go
struct {
    a int
}
```
